feat: Add gather & download logs button in troubleshooting

cdrage opened the pull request:
feat: Add gather & download logs button in troubleshooting

### What does this PR do?

* Adds a button to collect logs which is currently only hardcoded to
  retrieve the console logs
* Adds a button to download the logs as a zip file with all logs stored
  within as a .txt file
* Shows which files have been generated and what will be zipped.

### Screenshot / video of UI

<!-- If this PR is changing UI, please include
screenshots or screencasts showing the difference -->

https://github.com/containers/podman-desktop/assets/6422176/54bb8fd8-4f62-4ba5-9b50-4399f8c42c9a


### What issues does this PR fix or reference?

<!-- Include any related issues from Podman Desktop
repository (or from another issue tracker). -->

Closes https://github.com/containers/podman-desktop/issues/5048

### How to test this PR?

<!-- Please explain steps to reproduce -->

1. Go to troubleshooting
2. Click the collect logs & download logs buttons.

cdrage commented:
Ready for review!

Would like some input on how to improve the UI as well as other logs we should be collecting (Windows and Linux specifically).

In a later PR I would like to implement provider-related / extension-related details (for example, have an extension such as kind that will output the container logs).

deboer-tim reviewed (commented):
This review is going to jump around a lot, sorry in advance. :)

+1 to all the basic parts (zip, file dialog, etc) being there. It works!

Title: I know this as 'must gather' from the past 20 years and the name kinda sticks, but like 'onboarding' I don't know if the average user knows the term. Maybe something like "Gather Log Files" would be enough with an expanded description? e.g. "Bundle all log files into a zip for future (or offline?) problem determination".

From a core/UI standpoint I don't know when we'd ever want just console logs or just system logs, plus we'll have extension logs, etc. I think it's probably simpler if index.ts just exposes a 'getLogs', and if we ever need something more fine-grained it could have optional parameters in the future.

Going a step further I think it's nice it shows you a list of files that it's going to zip and uses mono-font to list them - but I think the usage pattern is always to click one, then the other. I'm wondering if we should even bother having a Collect Logs button and exposing it through the API, or just have a single button? Or maybe a single button and it just gives you a summary of the files that it included after.

We're a desktop app, so a "Download" button is a bit out of place. Maybe just "Save"?

jeffmaury commented on packages/main/src/plugin/troubleshooting.ts:
```diff
@@ -0,0 +1,94 @@
+    }
+  }
+
+  getConsoleLogs(consoleLogs: { logType: LogType; date: Date; message: string }[]): FileMap[] {
```
The date field is not used, is it on purpose ?

jeffmaury commented on packages/main/src/plugin/troubleshooting.ts:
```diff
@@ -0,0 +1,94 @@
+    if (os.platform() === 'darwin') {
+      logs = await this.getMacSystemLogs();
+    } else {
+      console.log('Unsupported platform for retrieving system logs');
```
Implement Windows and Linux

jeffmaury replied:
On Windows, the path is ~/AppData/Roaming/Podman Desktop/logs/podman-desktop.log

jeffmaury reviewed (dismissed):
Missing implementations

feloy commented on packages/main/src/plugin/troubleshooting.ts:
```diff
@@ -0,0 +1,110 @@
+  // TODO: Implement any other logs (Windows & Linux) we wish to retrieve in the future.
+  async getSystemLogs(): Promise<FileMap[]> {
+    let logs: FileMap[] = [];
+    if (os.platform() === 'darwin') {
+      logs = await this.getMacSystemLogs();
+    } else if (os.platform() === 'win32') {
+      logs = await this.getWindowsSystemLogs();
+    } else {
+      console.log('Unsupported platform for retrieving system logs');
```
I think we could "abstract" a log as either a file content or the output of a command. This way, we could have generic getFilesLogs() and getCommandLogs() functions, and in this if/elseif (or switch/case), we could just pass a list of files / commands.

feloy commented on packages/main/src/plugin/troubleshooting.ts:
```diff
@@ -0,0 +1,110 @@
+  }
+
+  // Retrieve the mac system log files from ~/Library/Logs/Podman Desktop
+  private async getMacSystemLogs(): Promise<FileMap[]> {
+    const logs: FileMap[] = [];
+    const logFiles = ['launchd-stdout', 'launchd-stderr'];
+    const logDir = `${os.homedir()}/Library/Logs/Podman Desktop`;
+    for (const file of logFiles) {
+      try {
+        const filePath = `${logDir}/${file}.log`;
+        const content = await fs.promises.readFile(filePath, 'utf8');
+        logs.push({ fileName: generateLogFileName('system-' + file), content });
+      } catch (error) {
+        console.error(`Error reading ${file}.log: `, error);
+      }
+    }
+    return logs;
+  }
+  // Retrieve the windows system log files from ~/AppData/Roaming/Podman Desktop/logs/podman-desktop.log
```
This would become a generic `getFileslog(files: string[])`
And we would pass for mac:
```
${os.homedir()}/Library/Logs/Podman Desktop/launchd-stdout
${os.homedir()}/Library/Logs/Podman Desktop/launchd-stderr
```

for windows:
```
${os.homedir()}/AppData/Roaming/Podman Desktop/logs/podman-desktop
```

cdrage commented:
Ready for another review 👍 / UI updated to use the new "tabbed" logs.

axel7083 commented on packages/renderer/src/lib/troubleshooting/TroubleshootingGatherLogs.svelte:
```diff
@@ -0,0 +1,86 @@
+      on:click="{() => retrieveLogs()}"
+      title="Collect logs for must gather tool"
+      icon="{faScroll}">Collect logs</Button>
+    <Button on:click="{() => saveLogsAsZip()}" title="Download .zip file" icon="{faDownload}" disabled="{!hasLogs}"
```
I feel like, this is redundant to click on collect, then click on save as zip.
We should be able to quickly do it in a single click

deboer-tim replied:
Similar feedback to what I was saying earlier, so I agree it should just be one button.

Maybe we can drop the 'console output' then too - if you care what files are included you can just open the zip before you give it to anyone.

With the new feedback tabs it also feels a little out of place as it's own tab (maybe should be on Logs tab? or maybe there's a different way to organize them) but we can always reorganize things later.

axel7083 commented on packages/main/src/plugin/troubleshooting.ts:
```diff
@@ -0,0 +1,95 @@
+
+  private async getFileSystemContent(filePath: string, logName: string): Promise<FileMap> {
+    try {
+      const content = await fs.promises.readFile(filePath, 'utf8');
```
I am getting an `Error: ENOENT: no such file or directory, open 'C:\Users\{pc}\AppData\Roaming\Podman Desktop\logs\podman-desktop.log'` 

Maybe we should check if the file exist before trying to open it ?


> I am on windows 11

cdrage replied:
Added a simple check to see if the file exists, if it doesnt, it'll skip it and output to log we were unable to find it.

axel7083 reviewed (commented):
Probably not part of this PR but would it be interesting to include this bundle while people are sharing feedback ? We could ask them if they want to include the logs, in case they are sharing a bug.

cdrage commented:
Ready for review now that 1.7.0 is out 👍

cdrage commented:
Ready for another review @deboer-tim @jeffmaury @benoitf @feloy 👍

deboer-tim commented:
Could we still merge this down to a single button and single exposed window. function?

cdrage commented:
> Could we still merge this down to a single button and single exposed window. function?

Done! Ready for another review 🎉

deboer-tim reviewed (changes requested):
One button: 👍🏼

I was expecting one button to also mean there is only one exposed function from main though too, e.g. something like window.saveLogsToZip(filename). All of the logic to gather console/system logs would be in here, and the UI only gets to say where to save it.

I guess implicit in the previous comment is that IMHO it would be best if the main package wasn't doing UI-things like opening the save dialog and we left this in renderer - i.e. GatherLog page would open the dialog to pick a location and then call the window function.

cdrage commented:
> One button: 👍🏼
> 
> I was expecting one button to also mean there is only one exposed function from main though too, e.g. something like window.saveLogsToZip(filename). All of the logic to gather console/system logs would be in here, and the UI only gets to say where to save it.
> 
> I guess implicit in the previous comment is that IMHO it would be best if the main package wasn't doing UI-things like opening the save dialog and we left this in renderer - i.e. GatherLog page would open the dialog to pick a location and then call the window function.

Done! Ready for another review. Put it down to one API call (that also gets the console logs via memoryLogs).

deboer-tim commented on packages/main/src/plugin/troubleshooting.ts:
```diff
@@ -0,0 +1,113 @@
+
+    const zip = new AdmZip();
+    fileMaps.forEach(fileMap => {
+      zip.addFile(`${fileMap.fileName}.txt`, Buffer.from(fileMap.content, 'utf8'));
```
Assumes every filename needs .txt, but maybe we want to include a .log or some other file in the future. Would be better if functions like getFileSystemContent() or generateLogFileName() set the filename they want and it just includes it here.

deboer-tim commented on packages/renderer/src/lib/troubleshooting/TroubleshootingGatherLogs.svelte:
```diff
@@ -0,0 +1,56 @@
+          <li>
+            <div class="flex flex-row align-middle items-center">
+              <div class="font-mono text-[10px] font-thin">
+                {log}.txt generated
```
Assumes that all files end in .txt and are generated. With the prior comment the correct filename would be here so you can drop the .txt, and I'd drop 'generated' too, since we don't know if the file was just copied in or generated. i.e. just use `{log}`.

deboer-tim reviewed (changes requested):
Thanks for all the changes, this is much simpler and cleaner to me now. I'd still appreciate a couple more minor tweaks as listed.

cdrage commented:
![Screenshot 2024-02-06 at 1 52 02 PM](https://github.com/containers/podman-desktop/assets/6422176/2729a997-4703-4330-ac3b-4450c32094e9)

Updated! I've also changed for it to retain the original file name extension, and if there is none, default to .txt

deboer-tim reviewed (approved):
Thanks for all the changes, LGTM.

benoitf commented on packages/main/src/plugin/troubleshooting.ts:
```diff
@@ -0,0 +1,118 @@
+      case 'win32':
+        return this.getLogsFromFiles(['podman-desktop'], `${os.homedir()}/AppData/Roaming/Podman Desktop/logs`);
+      default:
+        console.log('Unsupported platform for retrieving system logs');
```
or keep it as a comment

benoitf commented on packages/main/src/plugin/troubleshooting.ts:
```diff
@@ -0,0 +1,118 @@
+      const content = await fs.promises.readFile(filePath, 'utf8');
+      return { fileName: generateLogFileName(SYSTEM_FILENAME + '-' + logName), content };
+    } catch (error) {
+      console.error(`Error reading file from ${filePath}: `, error);
```
not sure we need the console as it's rethrown

benoitf commented on packages/main/src/plugin/troubleshooting.ts:
```diff
@@ -0,0 +1,118 @@
+
+        // Check if the file exists, if not, skip it.
+        if (!fs.existsSync(filePath)) {
+          console.log(`File ${filePath} does not exist, skipping`);
```
```suggestion

```

benoitf commented on packages/renderer/src/lib/troubleshooting/TroubleshootingGatherLogs.svelte:
```diff
@@ -0,0 +1,56 @@
+<script lang="ts">
+import { faFileLines, faScroll } from '@fortawesome/free-solid-svg-icons';
+import Button from '../ui/Button.svelte';
```
```suggestion
import Button from '/@/lib/ui/Button.svelte';
```

benoitf commented on packages/renderer/src/lib/troubleshooting/TroubleshootingGatherLogs.svelte:
```diff
@@ -0,0 +1,56 @@
+import { faFileLines, faScroll } from '@fortawesome/free-solid-svg-icons';
+import Button from '../ui/Button.svelte';
+import Fa from 'svelte-fa';
+import { generateLogFileName } from '../../../../main/src/plugin/troubleshooting';
```
I think here it's an issue as we import a class that is defined on the 'main' side and at some point it could not be anymore compliant with the client side

so probably need to be in a separate file

benoitf commented on packages/renderer/src/lib/troubleshooting/TroubleshootingGatherLogs.svelte:
```diff
@@ -0,0 +1,56 @@
+
+// Save files as a zip file (we first ask the user for the dialog, and then save the files to the filepath)
+async function saveLogsAsZip() {
+  const result = await window.saveFileDialog('Save Logs as .zip', generateLogFileName('podman-desktop', 'zip'));
```
another approach would be to expose a method 'troubleShootingGenerateLogFileName'

then from client side we ask the 'main side' instead of using a share function. I think it's too fragile like that

benoitf commented on packages/renderer/src/lib/troubleshooting/TroubleshootingGatherLogs.svelte:
```diff
@@ -0,0 +1,56 @@
+async function saveLogsAsZip() {
+  const result = await window.saveFileDialog('Save Logs as .zip', generateLogFileName('podman-desktop', 'zip'));
+  if (!result.canceled && result.filePath) {
+    const logFiles = await window.troubleshootingSaveLogs(result.filePath);
+    logs = logFiles;
```
maybe oneliner ?
```suggestion
logs = await window.troubleshootingSaveLogs(result.filePath);
```

cdrage commented:
@benoitf thank you for the review! implemented all the requested changes and it's ready for another 👍

cdrage commented:
Updated based on the changes, thank you for the reviews! Enabling auto-merge now 🎉

benoitf reviewed (commented):
I noticed that packages/renderer/src/lib/troubleshooting/TroubleshootingGatherLogs.svelte has no unit test
